detection_rules/navigator.py

```diff
@@ -0,0 +1,287 @@
+        layer[tactic][technique_id]['links'].append(self.rule_links_dict(rule))
+
+    def process_rule(self, rule: TOMLRule, tactic: str, technique_id: str):
+        self._update_all(rule, tactic, technique_id)
+        self._update_platforms(rule, tactic, technique_id)
+        self._update_indexes(rule, tactic, technique_id)
+        self._update_tags(rule, tactic, technique_id)
+
+    def process_rules(self):
+        """Adds rule to each applicable layer, including multi-layers."""
+        for rule in self.detection_rules:
+            threat = rule.contents.data.threat
+            if threat:
+                for entry in threat:
+                    tactic = entry.tactic.name.lower()
+                    if entry.technique:
+                        for technique_entry in entry.technique:
+                            technique_id = technique_entry.id
+                            self.process_rule(rule, tactic, technique_id)
+
+                            if technique_entry.subtechnique:
+                                for sub in technique_entry.subtechnique:
+                                    self.process_rule(rule, tactic, sub.id)
+
+    def build_navigator(self, layer_name: str, layer_key: Optional[str] = None) -> Navigator:
+        populated_techniques = []
+        layer = self.get_layer(layer_name, layer_key)
+        base_name = f'{layer_name}-{layer_key}' if layer_key else layer_name
+        name = f'Elastic-detection-rules-{base_name}'
+
+        for tactic, techniques in layer.items():
+            tactic_normalized = '-'.join(tactic.lower().split())
+            for technique_id, rules_data in techniques.items():
+                rules_data.update(tactic=tactic_normalized, techniqueID=technique_id)
+                techniques = Techniques.from_dict(rules_data)
+
+                populated_techniques.append(techniques.to_dict())
+
+        base_nav_obj = {
+            'name': name,
+            'techniques': populated_techniques,
+            'versions': {'attack': CURRENT_ATTACK_VERSION}
+        }
+        navigator = Navigator.from_dict(base_nav_obj)
+        return navigator
+
+    def build_all(self) -> List[Navigator]:
+        built = []
+
+        for layer_name, data in self.layers.items():
+            # this is a single layer
+            if 'defense evasion' in data:
+                built.append(self.build_navigator(layer_name))
+            else:
+                # multi layers
+                for layer_key, sub_data in data.items():
+                    built.append(self.build_navigator(layer_name, layer_key))
+
+        return built
+
+    @staticmethod
+    def _save(built: Navigator, directory: Path, verbose=True) -> Path:
+        path = directory.joinpath(built.name).with_suffix('.json')
+        path.write_text(json.dumps(built.to_dict(), indent=2))
+
+        if verbose:
+            print(f'saved: {path}')
+        return path
+
+    def save_layer(self,
```

Comment:
I'm not sure I am following, can you elaborate? Maybe the line breaks of the args?

Reply:
This is one of those cosmetic comments for consistency. The line length for most methods/functions appear to be 120.  We should probably decide to visually have multiple lines for the function signature or the minimum number of lines based on the preferred line length.

It's all really personal preference. I just noticed that it was slightly different. 

```
    def save_layer(self, layer_name: str, directory: Path, 
                   layer_key: Optional[str] = None, verbose=True) -> (Path, dict):

```

vs

```
    def save_layer(self,
                   layer_name: str,
                   directory: Path,
                   layer_key: Optional[str] = None,
                   verbose=True
                   ) -> (Path, dict):
```

Reply:
So I had always preferred your first option in this situation, but I have since leaning towards option B as of late because there have been more occurrences of odd breaks, mostly due to type hinting (also to stay more consistent with the stylings of other languages).

For now, I don't think we need to get too caught up on anything stylistic (though style guide is important for some cases)